# Aggregation and Grouping 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('darkgrid')

In [16]:
# load the data 
data = pd.read_excel("./Data/Global Superstore.xls")
data.head()

Row ID         Order ID Order Date  Ship Date     Ship Mode Customer ID  \
0   32298   CA-2012-124891 2012-07-31 2012-07-31      Same Day    RH-19495   
1   26341    IN-2013-77878 2013-02-05 2013-02-07  Second Class    JR-16210   
2   25330    IN-2013-71249 2013-10-17 2013-10-18   First Class    CR-12730   
3   13524  ES-2013-1579342 2013-01-28 2013-01-30   First Class    KM-16375   
4   47221     SG-2013-4320 2013-11-05 2013-11-06      Same Day     RH-9495   

      Customer Name      Segment           City            State  ...  \
0       Rick Hansen     Consumer  New York City         New York  ...   
1     Justin Ritter    Corporate     Wollongong  New South Wales  ...   
2      Craig Reiter     Consumer       Brisbane       Queensland  ...   
3  Katherine Murray  Home Office         Berlin           Berlin  ...   
4       Rick Hansen     Consumer          Dakar            Dakar  ...   

         Product ID    Category Sub-Category  \
0   TEC-AC-10003033  Technology  Accessories   
1   FUR-CH-10003950   Furniture       Chairs   
2   TEC-PH-10004664  Technology       Phones   
3   TEC-PH-10004583  Technology       Phones   
4  TEC-SHA-10000501  Technology      Copiers   

                                        Product Name     Sales Quantity  \
0  Plantronics CS510 - Over-the-Head monaural Wir...  2309.650        7   
1          Novimex Executive Leather Armchair, Black  3709.395        9   
2                  Nokia Smart Phone, with Caller ID  5175.171        9   
3                     Motorola Smart Phone, Cordless  2892.510        5   
4                     Sharp Wireless Fax, High-Speed  2832.960        8   

  Discount    Profit  Shipping Cost  Order Priority  
0      0.0  762.1845         933.57        Critical  
1      0.1 -288.7650         923.63        Critical  
2      0.1  919.9710         915.49          Medium  
3      0.1  -96.5400         910.16          Medium  
4      0.0  311.5200         903.04        Critical  

[5 rows x 24 columns]

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51290 entries, 0 to 51289
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Row ID          51290 non-null  int64         
 1   Order ID        51290 non-null  object        
 2   Order Date      51290 non-null  datetime64[ns]
 3   Ship Date       51290 non-null  datetime64[ns]
 4   Ship Mode       51290 non-null  object        
 5   Customer ID     51290 non-null  object        
 6   Customer Name   51290 non-null  object        
 7   Segment         51290 non-null  object        
 8   City            51290 non-null  object        
 9   State           51290 non-null  object        
 10  Country         51290 non-null  object        
 11  Postal Code     9994 non-null   float64       
 12  Market          51290 non-null  object        
 13  Region          51290 non-null  object        
 14  Product ID      51290 non-null  object        
 15  Ca

In [31]:
# I want to see the mean sales per category
data.groupby(by='Category')['Sales'].mean()

Category
Furniture          416.248905
Office Supplies    121.097120
Technology         467.858939
Name: Sales, dtype: float64

In [ ]:
for (method, group) in data.groupby(by='Category'):
    print(method,'-',group.shape)

Furniture - (9876, 24)
Office Supplies - (31273, 24)
Technology - (10141, 24)


### 🔎 What it’s doing:

- `data.groupby(by='Category')` → **Splits** the big `data` DataFrame into **smaller groups** based on unique values in the `Category` column.
  
- Then, **for each group**:
  - `method` will be the **name of the group** (like `'Furniture'`, `'Office Supplies'`, `'Technology'`).
  - `group` will be the **mini DataFrame** containing only rows that belong to that category.
  
- Inside the loop:
  - `group.shape` tells you **(number of rows, number of columns)** of that mini DataFrame.

- You print the category name (`method`) and the shape of that mini DataFrame.

---

### 📋 Your output:

| Category        | Shape        | Meaning |
|:----------------|:-------------|:--------|
| Furniture        | (9876, 24)    | 9876 rows, 24 columns in Furniture category |
| Office Supplies | (31273, 24)  | 31273 rows, 24 columns in Office Supplies category |
| Technology      | (10141, 24)  | 10141 rows, 24 columns in Technology category |

---

# 🎯 In simple words:
You **looped through each Category** and printed:
- the category name
- and how many records (rows) and fields (columns) it has.

---

# 🧠 Visually:

```
Big DataFrame (total rows = 51290, columns = 24)
    ├── 'Furniture' group → 9876 rows
    ├── 'Office Supplies' group → 31273 rows
    └── 'Technology' group → 10141 rows
```

Each group is just a **subset** of the original data, grouped by `Category`.

---

# 🔥 Bonus Tip:

If you ever just want the **sizes** without looping manually, you can do:

```python
data.groupby('Category').size()
```
This will directly give you:

```
Category
Furniture          9876
Office Supplies   31273
Technology        10141
dtype: int64
```

---

## 🎯 What are **dispatch methods** in pandas?

**Dispatch methods** are **normal DataFrame or Series methods** that **pandas automatically sends ("dispatches") to each group** when you use `groupby`.

✅ In other words:  
When you do something like `.sum()`, `.mean()`, `.min()`, `.max()`, `.count()`, etc. **after `groupby`**, pandas will automatically apply those methods **to each group separately**, **without you needing to write a custom function**.

---

# 📦 Examples of dispatch methods:
- `.sum()`
- `.mean()`
- `.min()`
- `.max()`
- `.count()`
- `.first()`
- `.last()`
- `.std()`
- `.var()`
- `.median()`
- `.nunique()`
- etc.

**They are called "dispatch methods" because pandas "dispatches" the work to each group automatically.**

---

### 🛠 How dispatch methods behave on `groupby`?

When you use a dispatch method after `groupby`:

1. **Pandas splits the DataFrame** into groups.
2. **It runs the method** separately on **each group**.
3. **It combines** the results back into a single new DataFrame or Series.

---

#### 📚 Let's see an example:

```python
import pandas as pd

data = pd.DataFrame({
    'Category': ['Furniture', 'Furniture', 'Technology', 'Office Supplies', 'Technology', 'Office Supplies'],
    'Sales': [200, 300, 400, 150, 600, 250]
})

# Group by Category and find the sum of Sales
grouped_sales = data.groupby('Category')['Sales'].sum()
print(grouped_sales)
```

🔵 Output:

```
Category
Furniture           500
Office Supplies     400
Technology         1000
Name: Sales, dtype: int64
```

✅ `.sum()` was dispatched automatically to each group!

- **Furniture** → 200 + 300 = 500
- **Office Supplies** → 150 + 250 = 400
- **Technology** → 400 + 600 = 1000

---

### 📢 Summary:  
| Feature | What it means |
|:---|:---|
| **Dispatch Methods** | Normal DataFrame/Series methods that pandas automatically runs on each group |
| **Behavior** | Splits → Applies the method on each group → Combines the results |
| **Examples** | `.sum()`, `.mean()`, `.count()`, `.min()`, `.max()`, `.std()`, etc. |

---

### 🚀 Quick difference:

| Concept | Behavior |
|:---|:---|
| `.apply()` | You must manually tell pandas what custom logic to apply on each group. |
| **Dispatch method** (e.g., `.sum()`) | Pandas **already knows** what to do — no custom function needed. |

---

### ⚡ One-liner to remember:

> **If pandas already knows what you want (sum, mean, count, etc.), use dispatch methods. Otherwise, use `apply()`.**

---

In [44]:
data.groupby(by='Category')['Sales'].describe()

count        mean         std    min      25%     50%  \
Category                                                                   
Furniture         9876.0  416.248905  553.066417  1.892   91.026  220.20   
Office Supplies  31273.0  121.097120  299.321405  0.444   20.340   46.38   
Technology       10141.0  467.858939  708.972701  0.990  117.150  260.34   

                     75%        max  
Category                             
Furniture        509.265   5759.964  
Office Supplies  105.240   9892.740  
Technology       554.850  22638.480

## 🎯 What `.agg()` can take?

When you use `.agg()` after a `groupby` (or even directly on a DataFrame), you can pass in:

| What you pass | Example | What it means |
|:---|:---|:---|
| **Single function name as string** | `'mean'`, `'sum'`, `'min'`, `'max'`, `'count'` | Apply one function |
| **List of function names** | `['mean', 'sum']` | Apply multiple built-in functions |
| **Custom function** | `lambda x: x.max() - x.min()` | Your own function |
| **List of custom + built-in functions** | `['mean', lambda x: x.max() - x.min()]` | Mix and match |
| **Dictionary** | `{'Marks': ['mean', 'max'], 'Age': 'min'}` | Apply specific functions to specific columns |

---

## 🛠️ Examples

### 1. **Single built-in function**
```python
df.groupby('Class')['Marks'].agg('mean')
```
👉 Find average marks per class.

---

### 2. **List of multiple built-in functions**
```python
df.groupby('Class')['Marks'].agg(['mean', 'sum', 'max'])
```
👉 Find mean, sum, and max marks per class.

---

### 3. **Custom function**
```python
df.groupby('Class')['Marks'].agg(lambda x: x.max() - x.min())
```
👉 Find the range (max - min) marks per class.

---

### 4. **Mixing built-in and custom functions**
```python
df.groupby('Class')['Marks'].agg(['mean', lambda x: x.max() - x.min()])
```
👉 Get mean and range together.

---

### 5. **Using dictionary (different columns different functions)**

Suppose you have more columns:
```python
data = {
    'Class': ['A', 'A', 'B', 'B', 'B', 'C'],
    'Marks': [85, 90, 78, 82, 89, 95],
    'Age': [14, 15, 14, 15, 14, 16]
}

df = pd.DataFrame(data)
```

You can do:

```python
df.groupby('Class').agg({
    'Marks': ['mean', 'max'],
    'Age': 'min'
})
```

👉 Here:
- For `Marks`, it finds mean and max.
- For `Age`, it finds minimum.

---

## 📋 Built-in function names you can pass (examples):

| Function | Meaning |
|:---|:---|
| `'mean'` | Average |
| `'sum'` | Total |
| `'min'` | Minimum |
| `'max'` | Maximum |
| `'count'` | Number of non-null entries |
| `'std'` | Standard deviation |
| `'var'` | Variance |
| `'median'` | Median |
| `'first'`, `'last'` | First/last value |
| `'nunique'` | Number of unique values |

*(and a few others, but these are most common)*

---

## 🔥 In short:
- `.agg()` is like a **Swiss army knife** 🛠️ — you can customize **what** and **how many things** you want to calculate.
- You can pass strings, lists, functions, and even dictionaries!

---

## 🔥 Quick thumb rule:

> **"If it's simple math/statistics ➔ use `agg()`."**  
> **"If it's complex decisions/transformations ➔ use `apply()`."**

---


In [ ]:
# Say, I want to categorize sales for each category as Above Avergae abd Below Average based on whether they are greater than average sales for that category

# Define a custom function to categorize sales within each group
def categorize_sales(group):
    # Calculate the average sale for the current group (each 'Category')
    avg_sale = group['Sales'].mean()
    
    # For each Sale in the group, label it as 'Above Average' or 'Below Average'
    # based on whether it is greater than the group's average
    return group['Sales'].apply(
        lambda x: 'Above Average' if x > avg_sale else 'Below Average'
    )

# Group the data by 'Category'
# Then apply the 'categorize_sales' function to each group
# 'include_groups=False' ensures that the 'Category' column is NOT included inside each group (future-proof)
sales_category = data.groupby('Category').apply(categorize_sales, include_groups=False)

# View the resulting Series
sales_category


Category         
Furniture   1        Above Average
            6        Above Average
            7        Above Average
            9        Above Average
            11       Above Average
                         ...      
Technology  51151    Below Average
            51178    Below Average
            51219    Below Average
            51237    Below Average
            51259    Below Average
Name: Sales, Length: 51290, dtype: object

In [30]:
# Combine with original DataFrame
df_clean = data.copy(deep=True)
df_clean['Sales_Category'] = sales_category.values  # .values to align the index
df_clean.head()

Row ID         Order ID Order Date  Ship Date     Ship Mode Customer ID  \
0   32298   CA-2012-124891 2012-07-31 2012-07-31      Same Day    RH-19495   
1   26341    IN-2013-77878 2013-02-05 2013-02-07  Second Class    JR-16210   
2   25330    IN-2013-71249 2013-10-17 2013-10-18   First Class    CR-12730   
3   13524  ES-2013-1579342 2013-01-28 2013-01-30   First Class    KM-16375   
4   47221     SG-2013-4320 2013-11-05 2013-11-06      Same Day     RH-9495   

      Customer Name      Segment           City            State  ...  \
0       Rick Hansen     Consumer  New York City         New York  ...   
1     Justin Ritter    Corporate     Wollongong  New South Wales  ...   
2      Craig Reiter     Consumer       Brisbane       Queensland  ...   
3  Katherine Murray  Home Office         Berlin           Berlin  ...   
4       Rick Hansen     Consumer          Dakar            Dakar  ...   

     Category  Sub-Category  \
0  Technology   Accessories   
1   Furniture        Chairs   
2  Technology        Phones   
3  Technology        Phones   
4  Technology       Copiers   

                                        Product Name     Sales Quantity  \
0  Plantronics CS510 - Over-the-Head monaural Wir...  2309.650        7   
1          Novimex Executive Leather Armchair, Black  3709.395        9   
2                  Nokia Smart Phone, with Caller ID  5175.171        9   
3                     Motorola Smart Phone, Cordless  2892.510        5   
4                     Sharp Wireless Fax, High-Speed  2832.960        8   

  Discount    Profit Shipping Cost  Order Priority  Sales_Category  
0      0.0  762.1845        933.57        Critical   Above Average  
1      0.1 -288.7650        923.63        Critical   Above Average  
2      0.1  919.9710        915.49          Medium   Above Average  
3      0.1  -96.5400        910.16          Medium   Above Average  
4      0.0  311.5200        903.04        Critical   Above Average  

[5 rows x 25 columns]